engines/e_padlock.c

```diff
@@ -452,11 +560,108 @@ padlock_ctr_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out_arg,
     return 1;
 }
 
-#  define EVP_CIPHER_block_size_ECB       AES_BLOCK_SIZE
-#  define EVP_CIPHER_block_size_CBC       AES_BLOCK_SIZE
-#  define EVP_CIPHER_block_size_OFB       1
-#  define EVP_CIPHER_block_size_CFB       1
-#  define EVP_CIPHER_block_size_CTR       1
+static int
+gmi_sm4_ecb_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out_arg,
+                   const unsigned char *in_arg, size_t nbytes)
+{
+    struct gmi_cipher_data *cdata = ALIGNED_CIPHER_DATA_GMI(ctx);
+    
+    gmi_sm4_encrypt(out_arg, in_arg, cdata, nbytes);
+    return 1;
+}
+
+static int
+gmi_sm4_cbc_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out_arg,
+                   const unsigned char *in_arg, size_t nbytes)
+{
+    struct gmi_cipher_data *cdata = ALIGNED_CIPHER_DATA_GMI(ctx);
+    
+    memcpy(cdata->iv, EVP_CIPHER_CTX_iv(ctx), SM4_BLOCK_SIZE);
+    gmi_sm4_encrypt(out_arg, in_arg, cdata, nbytes);
+    memcpy(EVP_CIPHER_CTX_iv_noconst(ctx), cdata->iv, SM4_BLOCK_SIZE);   
+    return 1;
+}
+
+static int
+gmi_sm4_ctr_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out_arg,
+                   const unsigned char *in_arg, size_t nbytes)
+{
+    struct gmi_cipher_data *cdata = ALIGNED_CIPHER_DATA_GMI(ctx);
+    unsigned int num = EVP_CIPHER_CTX_num(ctx);
+    
+    if (f_zxc == 1) {
+        unsigned char * buf = EVP_CIPHER_CTX_buf_noconst(ctx);
+        CRYPTO_ctr128_encrypt(in_arg, out_arg, nbytes,
+                              cdata->ks.rd_key, EVP_CIPHER_CTX_iv_noconst(ctx), buf, &num,
+                              (block128_f) gmi_sm4_ecb_enc);
+       EVP_CIPHER_CTX_set_num(ctx, num);
+    } else {
+        memcpy(cdata->iv, EVP_CIPHER_CTX_iv(ctx), SM4_BLOCK_SIZE);
+        if (nbytes % SM4_BLOCK_SIZE) {        
+            nbytes = SM4_BLOCK_SIZE - nbytes % SM4_BLOCK_SIZE + nbytes;    
+        }
+        gmi_sm4_encrypt(out_arg, in_arg, cdata, nbytes);
+        memcpy(EVP_CIPHER_CTX_iv_noconst(ctx), cdata->iv, SM4_BLOCK_SIZE);
+        EVP_CIPHER_CTX_set_num(ctx, num);
```

Comment:
This is wrong, you have to update "num" the number of bytes on the remaining block,
see the AES code for an example or CRYPTO_ctr128_encrypt for how to do that.
maybe we are missing some known answer tests for SM4?